File: 2023/UICPC DIV1/UICPC Round #05/solutions/openpitmining/submissions/accepted/openpitmining-zf.py
#!/usr/bin/python3

from collections import deque

INFTY = 10000000000000

edges = []
graph = dict()

def addEdge(u, v, c):
    if u not in graph: graph[u] = set()
    graph[u].add(len(edges))
    edges.append((u, v, c))

    if v not in graph: graph[v] = set()
    graph[v].add(len(edges))
    edges.append((v, u, 0))


def augment(s, t):
    global edges

    path = {s: -1}
    q = deque([s])

    while q:
        cur = q.popleft()
        for i in graph[cur]:
            nxt = edges[i][1]
            if edges[i][2] > 0 and nxt not in path:
                path[nxt] = i
                q.append(nxt)

    if t not in path: return 0

    aug = INFTY
    tt = t
    while tt != s:
        i = path[tt]
        tt = edges[i][0]
        aug = min(aug, edges[i][2])

    tt = t
    while tt != s:
        i = path[tt]
        tt = edges[i][0]
        edges[i] = (edges[i][0], edges[i][1], edges[i][2]-aug)
        edges[i^1] = (edges[i^1][0], edges[i^1][1], edges[i^1][2]+aug)

    return aug

def maxFlow(s, t):
    flow = 0
    while True:
        aug = augment(s, t)
        if aug == 0: return flow
        flow += aug

n = int(input())
totVal = 0
for i in range(1, n+1):
    line = list(map(int, input().split()))
    totVal += line[0]
    addEdge(0, i, line[0])
    addEdge(i, n+1, line[1])
    for obs in line[3:]:
        addEdge(obs, i, INFTY)

print(totVal-maxFlow(0, n+1))

File: 2023/UICPC DIV1/UICPC Round #05/solutions/openpitmining/submissions/accepted/openpitmining-zf.cpp
#include <iostream>
#include <vector>
#include <queue>

#define INFTY 10000000000000ll

using namespace std;

typedef long long ll;

struct edge {
    ll u, v, c;
    edge(ll uu = 0, ll vv = 1, ll cc = 0) :
        u(uu), v(vv), c(cc) {}
};

vector<edge> edges;
vector<vector<int>> g;
vector<int> path;

void addEdge(int u, int v, ll c) {
    g[u].push_back(edges.size());
    edges.push_back(edge(u, v, c));
    g[v].push_back(edges.size());
    edges.push_back(edge(v, u, 0));
}

ll augment(int s, int t) {
    path.assign(g.size(), -1);
    queue<int> q;

    q.push(s);
    path[s] = -2;

    int its = 0;

    while (!q.empty()) {
        int cur = q.front();
        q.pop();
        ++its;
        for (auto i : g[cur]) {
            int nxt = edges[i].v;
            if (edges[i].c && path[nxt] == -1) {
                path[nxt] = i;
                q.push(nxt);
            }
        }
    }

    if (path[t] == -1) return 0;

    ll aug = INFTY;
    int tt = t;
    while (tt != s) {
        int i = path[tt];
        tt = edges[i].u;
        aug = min(aug, edges[i].c);
    }

    tt = t;
    while (tt != s) {
        int i = path[tt];
        tt = edges[i].u;
        edges[i].c -= aug;
        edges[i^1].c += aug;
    }

    return aug;
}

ll maxFlow(int s, int t) {
    ll flow = 0, aug;

    do {
        flow += (aug = augment(s, t));
    } while (aug);

    return flow;
}

int main() {
    int n;
    cin >> n;

    ll totVal = 0;
    g.resize(n+2);
    for (int i = 1; i <= n; ++i) {
        ll v, c;
        int m;

        cin >> v >> c >> m;
        totVal += v;
        addEdge(0, i, v);
        addEdge(i, n+1, c);
        while (m--) {
            int obs;
            cin >> obs;
            addEdge(obs, i, INFTY);
        }
    }

    cout << totVal - maxFlow(0, n+1) << endl;
    // for (int i = 1; i < edges.size(); i += 2)
    //     if (edges[i].c)
    //     cout << edges[i^1].u << ' ' << edges[i^1].v << ' ' << edges[i].c << endl;

    return 0;
}
